### DISCUSSION TITLE
amxmlc exits with "Error: null" using Java 8, seems to work with Java 7

### DISCUSSION THREAD

**Discussion by rnmixon** [airsdk/Adobe-Runtime-Support#2470] [General]
I am trying to use a script to create a three step build of a complex AIR application: compile to .swf, generate AIR independent .air file and then create a bundle with embedded runtime. The eventual target will be Linux, but I am trying to get this to run on Windows first since we are able to achieve the result interactively using Flash Builder 4.7.

The script works fine for a very simple AIR "CountDown" application, but when we substitute in large AIR application it fails. To even get this far I had to update the amxmlc.bat file to increase the Java heap size.

After a bit more reading I saw that several years ago there was a problem getting amxmlc to run when Java 8 first came out, so I substituted an archived version of Java 7 (JRE 1.7.0_80-b15). This seems to create the .swf file - but that's a really obsolete EOL version of Java - the Harman specs say to use Java 8.

The amxlmc just finishes with the string "Error: null". Any idea on how to get this to work with Java 8 or at least turn on a trace that shows where things go bad?

I'm including the output from running the batch file - first using Java 7 and then Java 8. I've instrumented the amxmlc.bat file to dump the version of java that's being used just before it runs the actual mxmlc.jar file.

**Output using Java 7**

```
C:\Users\rnmixon\Adobe Flash Builder 4.7\n2kview>..\build.cmd N2KViewBlackBox n2kview src\N2KViewBlackBox src\N2KViewBlackBox
AIR_HOME: C:\Users\rnmixon\Adobe Flash Builder 4.7\n2kcommon\sdks\Flex4.16.0withAIR33.1
adt: C:\Users\rnmixon\Adobe Flash Builder 4.7\n2kcommon\sdks\Flex4.16.0withAIR33.1\bin\adt
PROG_NAME: N2KViewBlackBox
PROG_FOLDER_NAME: n2kview
PROG_MXML_NAME: src\N2KViewBlackBox
PROG_XML_NAME: src\N2KViewBlackBox
SRC_DIR: C:\Users\rnmixon\Adobe Flash Builder 4.7\n2kview
Complete
AIR compiled "N2KViewBlackBox.swf" NOT found, continuing
AIR intermediate "N2KViewBlackBox.airi" NOT found, continuing.
Device independent Mac AIR file "N2KViewBlackBox.air" NOT found, continuing.
PC captive runtime bundle folder N2KViewBlackBox_bundle does NOT exist, continuing.
Package folder C:\Users\rnmixon\Adobe Flash Builder 4.7\n2kview\package already exists, cleaning.
Package folder "C:\Users\rnmixon\Adobe Flash Builder 4.7\n2kview\package" cleaned.
Creating AIR SWF file "N2KViewBlackBox.swf", ...
Loading extra config amxmlc-options.xml

C:\Users\rnmixon\Adobe Flash Builder 4.7\n2kview>amxmlc -load-config+=amxmlc-options.xml -output "N2KViewBlackBox.swf" -warnings "src\N2KViewBlackBox.mxml"
E:\tools\jre1.7.0_80
java version "1.7.0_80"
Java(TM) SE Runtime Environment (build 1.7.0_80-b15)
Java HotSpot(TM) 64-Bit Server VM (build 24.80-b11, mixed mode)
Loading configuration file C:\Users\rnmixon\Adobe Flash Builder 4.7\n2kcommon\sdks\Flex4.16.0withAIR33.1\frameworks\air-config.xml
Loading configuration file C:\Users\rnmixon\Adobe Flash Builder 4.7\n2kview\amxmlc-options.xml
C:\Users\rnmixon\Adobe Flash Builder 4.7\n2kcommon\sdks\Flex4.16.0withAIR33.1\frameworks\libs\spark.swc$defaults.css(162): Warning: The style 'backgroundColor' is only supported by type 'spark.components.Panel' with the theme(s) 'spark, mobile'.

C:\Users\rnmixon\Adobe Flash Builder 4.7\n2kcommon\sdks\Flex4.16.0withAIR33.1\frameworks\libs\spark.swc$defaults.css(234): Warning: The style 'cornerRadius' is only supported by type 'spark.components.TabBar' with the theme(s) 'spark'.

C:\Users\rnmixon\Adobe Flash Builder 4.7\n2kview\N2KViewBlackBox.swf (39258972 bytes)

C:\Users\rnmixon\Adobe Flash Builder 4.7\n2kview>dir N2KViewBlackBox.swf
02/16/2023  04:13 PM        39,258,972 N2KViewBlackBox.swf
C:\Users\rnmixon\Adobe Flash Builder 4.7\n2kview>
```

**Output using Java 8**
```
C:\Users\rnmixon\Adobe Flash Builder 4.7\n2kview>..\build.cmd N2KViewBlackBox n2kview src\N2KViewBlackBox src\N2KViewBlackBox
AIR_HOME: C:\Users\rnmixon\Adobe Flash Builder 4.7\n2kcommon\sdks\Flex4.16.0withAIR33.1
adt: C:\Users\rnmixon\Adobe Flash Builder 4.7\n2kcommon\sdks\Flex4.16.0withAIR33.1\bin\adt
PROG_NAME: N2KViewBlackBox
PROG_FOLDER_NAME: n2kview
PROG_MXML_NAME: src\N2KViewBlackBox
PROG_XML_NAME: src\N2KViewBlackBox
SRC_DIR: C:\Users\rnmixon\Adobe Flash Builder 4.7\n2kview
Complete
Deleted "N2KViewBlackBox.swf"
AIR intermediate "N2KViewBlackBox.airi" NOT found, continuing.
Device independent Mac AIR file "N2KViewBlackBox.air" NOT found, continuing.
PC captive runtime bundle folder N2KViewBlackBox_bundle does NOT exist, continuing.
Package folder C:\Users\rnmixon\Adobe Flash Builder 4.7\n2kview\package already exists, cleaning.
Package folder "C:\Users\rnmixon\Adobe Flash Builder 4.7\n2kview\package" cleaned.
Creating AIR SWF file "N2KViewBlackBox.swf", ...
Loading extra config amxmlc-options.xml

C:\Users\rnmixon\Adobe Flash Builder 4.7\n2kview>amxmlc -load-config+=amxmlc-options.xml -output "N2KViewBlackBox.swf" -warnings "src\N2KViewBlackBox.mxml"
C:\Program Files\BellSoft\LibericaJRE-8-Full
openjdk version "1.8.0_362"
OpenJDK Runtime Environment (build 1.8.0_362-b09)
OpenJDK 64-Bit Server VM (build 25.362-b09, mixed mode)
OpenJDK 64-Bit Server VM warning: ignoring option MaxPermSize=256m; support was removed in 8.0
OpenJDK 64-Bit Server VM warning: ignoring option PermSize=256m; support was removed in 8.0
Loading configuration file C:\Users\rnmixon\Adobe Flash Builder 4.7\n2kcommon\sdks\Flex4.16.0withAIR33.1\frameworks\air-config.xml
Loading configuration file C:\Users\rnmixon\Adobe Flash Builder 4.7\n2kview\amxmlc-options.xml
Error: null

C:\Users\rnmixon\Adobe Flash Builder 4.7\n2kview>dir N2KViewBlackBox.swf
File Not Found
C:\Users\rnmixon\Adobe Flash Builder 4.7\n2kview>
```

Thank you much - Richard

**Comment by ajwfrost** [comment]
(I just moved the triple-backtick mark in your post to make the formatting work..)

The compiler you're using there is the mxmlc.jar from the Flex SDK.. it's not the same as the one that we can build and that is provided as part of the AIR SDK "with new compiler". Given that the xml config files are the same between these, and it works for a basic SWF, I would think the only options are:
a) grab the Flex SDK source code and build the compiler with additional debug output (afaik there's no "verbose" option..)
b) gradually bridge the gap between the simple app and your complex one, to see if you can spot where things start failing e.g. with the config file, or with ANEs, or similar.

But actually, and given that it is the Flex compiler you're using, I wouldn't see any harm in using Java 7 for the "SWF compilation" step, as long as you moved to Java 8 or 11 for the ADT packaging process.

thanks

**Reply by rnmixon** [reply]
Andrew,
Just so I'm clear ... it  seems you are saying that the new compiler included with the SDK "Full AIR SDK with new ActionScript Compiler"  will not work with a Flex project.  Too bad.

Moving on, for now I'm running amxmlc.bat under Java 7 and the two subsequent adt.bat steps under Java 8. When I try to run adt to convert the .swf file to a .airi file I get the following errors:

```
C:\Users\rnmixon\Adobe Flash Builder 4.7\n2kview>call "E:\tools\apache-flex-sdk\bin\adt" -prepare N2KViewBlackBox.airi src\N2KViewBlackBox-app.xml N2KViewBlackBox.swf

C:\Users\rnmixon\Adobe Flash Builder 4.7\n2kview>java -Xms512m -Xmx3072m -XX:MaxPermSize=256m -XX:PermSize=256m -jar "E:\tools\apache-flex-sdk\bin\\..\lib\adt.jar" -prepare N2KViewBlackBox.airi src\N2KViewBlackBox-app.xml N2KViewBlackBox.swf
OpenJDK 64-Bit Server VM warning: ignoring option MaxPermSize=256m; support was removed in 8.0
OpenJDK 64-Bit Server VM warning: ignoring option PermSize=256m; support was removed in 8.0
C:\Users\rnmixon\Adobe Flash Builder 4.7\n2kview\src\N2KViewBlackBox-app.xml: error 303: Icon icons/Icon48.png is missing from package
C:\Users\rnmixon\Adobe Flash Builder 4.7\n2kview\src\N2KViewBlackBox-app.xml: error 303: Icon icons/Icon128.png is missing from package
C:\Users\rnmixon\Adobe Flash Builder 4.7\n2kview\src\N2KViewBlackBox-app.xml: error 303: Icon icons/Icon32.png is missing from package

C:\Users\rnmixon\Adobe Flash Builder 4.7\n2kview>
```
The three referenced icon files are in subfolder src/icons of the main "n2kview" folder. I'm assuming they should have been moved into the .swf file - or is that not the case?

For that matter is my adt command line correct? I'm referencing both the associate "-app.xml" file and the .swf file.

Thank you much - Richard

**Reply by ajwfrost** [reply]
>  it seems you are saying that the new compiler included with the SDK "Full AIR SDK with new ActionScript Compiler" will not work with a Flex project. 

Yes exactly - there are two variants of the AIR SDK, one of them doesn't include the compiler so that it can be copied over the top of a Flex SDK to provide the updated runtime and AS3 APIs, but still have the mxml support required by Flex. The "new" AS3 compiler doesn't proper support for everything within Flex, I'm not sure why Adobe didn't continue with it to add in the capabilities but perhaps by that point the Flex stuff was all owned by Apache..

In terms of the conversion to .airi, the complaint is that the icon files have not been included in the file list for packaging into the .airi file. Your app descriptor file references them as `icons/Icon48.png` etc, so they need to be added to the command line, perhaps if you just end that with `-C src icons` it would pick them up properly? The `-C src` essentially changes the folder that it looks at to the `src` one, and then the `icons` mention would mean it picks up everything within the `icons` folder, with the paths starting with `icons` as well.

They don't get moved into the SWF at all, they just get zipped in with the rest of the files. So yes, the rest of the adt command looks fine, you give it the XML file first and then all other necessary files should follow that.

thanks

**Reply by rnmixon** [reply]
The "-C src icons" parameter did the trick. I was able to go through all three steps and create the bundle. 

It comes up with a black screen, but I probably just need to get the application developer looped in to figure that out.

Thank you much - Richard
